File: Arrays/valid-anagram/python/valid_anagram.py
class Solution:
    @staticmethod
    def isAnagram(s: str, t:str)->bool:
        if len(s) != len(t):
            return False
        frequency, TARGET = [0] * 26, [0] * 26
        
        #Zip is faster and more efficient than range, it also skips pointer index lookups 
        #s[i] or t[i]
        for (s_t, t_t) in zip(s, t):
            frequency[ord(s_t) - 97]+=1
            frequency[ord(t_t) - 97]-=1
            # Instead of using all, this is way faster internally
        return frequency == TARGET
            
        

if __name__ == "__main__":
    test_cases = [
        ("anagram", "nagaram"),   # → True, classic example
        ("rat", "car"),           # → False, different letters
        ("a", "a"),               # → True, single char
        ("a", "b"),               # → False, single char mismatch
        ("aa", "a"),              # → False, different lengths
        ("listen", "silent"),     # → True, classic anagram pair
    ]

    for s, t in test_cases:
        print(f"Input:       s={s!r}, t={t!r}")
        print(f"Output:      {Solution().isAnagram(s, t)}")
        print("-" * 35)

File: Arrays/valid-anagram/ts/validAnagram.ts
export {};

//comparing frequency table
function _isAnagram(s: string, t: string): boolean {
  if (s.length != t.length) {
    return false;
  }
  let s_map = new Map<string, number>();
  let t_map = new Map<string, number>();
  for (let i = 0; i < s.length; i++) {
    s_map.set(s[i], (s_map.get(s[i]) || 0) + 1);
    t_map.set(t[i], (t_map.get(t[i]) || 0) + 1);
  }

  if (s_map.size !== t_map.size) {
    return false;
  }

  for (const [key, value] of s_map) {
    if (t_map.get(key) != value) {
      return false;
    }
  }
  return true;
}

//Array fill method
function isAnagram(s: string, t: string): boolean {
  if (s.length != t.length) {
    return false;
  }
  let frequency = Array(26).fill(0);
  for (let i = 0; i < s.length; i++) {
    let sVal = s.charCodeAt(i) - 97;
    frequency[sVal]++
    let tVal = t.charCodeAt(i) - 97;
    frequency[tVal]--
  }
  return frequency.every((val) => val == 0)
}

const testCases: [string, string][] = [
  ["anagram", "nagaram"],
  ["rat", "car"],
  ["a", "a"],
  ["a", "b"],
  ["aa", "a"],
  ["listen", "silent"],
];

for (const [s, t] of testCases) {
  console.log(`Input:       s=${JSON.stringify(s)}, t=${JSON.stringify(t)}`);
  console.log(`Output:      ${isAnagram(s, t)}`);
  console.log("-".repeat(35));
}
